task_sdk/src/airflow/sdk/execution_time/task_runner.py

```diff
@@ -535,10 +556,16 @@ def run(ti: RuntimeTaskInstance, log: Logger):
             state=TerminalTIState.FAILED,
             end_date=datetime.now(tz=timezone.utc),
         )
+        get_listener_manager().hook.on_task_instance_failed(
+            previous_state=TaskInstanceState.RUNNING, task_instance=ti
+        )
         # TODO: Run task failure callbacks here
     except BaseException:
         log.exception("Task failed with exception")
         # TODO: Run task failure callbacks here
+        get_listener_manager().hook.on_task_instance_failed(
+            previous_state=TaskInstanceState.RUNNING, task_instance=ti
+        )
```

Comment:
Should we move all the calls to `finalize` and we can change sig of finalize to `state, log` etc?

That might be a better design than putting this on all `except` blocks.

Reply:
iirc @ashb had Listeners in mind for the `finalize` function.

Reply:
Yeah listeners, certainly post-execute ones in finalize please.

Reply:
Done.